### DISCUSSION TITLE
Can't execute and interpolated string as a command

### DISCUSSION THREAD

**Discussion by LinArcX** [nushell/nushell#7794] [Q&A] [answered]
Hello.

I want to create a string containing this command: `clang++ --help`

In bash, it shows the help page of the clang++ compiler.

What I've done so far in the Nushell is that I created a file called `show_help.nu`:
```
let compiler = "clang++"
let help_param = "--help"
^$"($compiler) ($help_param)"
```

But if I run it, I'll get this error message:
```
Error: nu::shell::external_command (link)

  × External command failed
   ╭─[build.nu:5:1]
 5 │ let help_param = "--help"
 6 │ ^$"($compiler) ($help_param)"
   ·  ──────────────┬─────────────
   ·                ╰── executable was not found
 7 │
   ╰────
  help: No such file or directory (os error 2)
```

Any idea to fix that?

**Comment by LinArcX** [comment]
Ok, I found a solution(Not sure if it's the best solution or not, please tell me guys):
```
let help_param = "--help"
let full_command = $compiler + " " + $help_param
nu -c $full_command
```

**Comment by fdncred** [comment, marked as answer]
This works too
```
 ^$"($compiler)" $help_param
```
I think the problem was that it was looking for 1 command named `"clang++ --help"` versus a command with a parameter.

**Comment by yerlaser** [comment]
This issue exhibits itself in many more situations.
```
let fl1 = '-F'
let fl2 = '-A'
let fls = $'($fl1) ($fl2)'
^ls $fls
```
Gives an error, while running:
`^ls -F -A`
works fine.
I know I could run it with `nu -c` but that spawns a new process which is not always desired.

Is there a proper way of constructing options for external commands?

**Reply by fdncred** [reply]
You want to keep the arguments separate from the command and use string interpolation on the external command.
```
^$"ls" ($fl1) ($fl2)
```

**Reply by yerlaser** [reply]
Thanks!
But, for my use case, it still won't work.
I need to construct a list of same flags in a `for` loop like this: `-p 8080:8080 -p 8081:8081 -p 8082:8082`. There can be variable count of these flags that's why they need to be constructed in a loop and be in one single string accumulator variable.
Then when I run:
`docker run -it --rm $ports ubuntu`
this string variable gets passed as a single, I believe, quoted string and thus docker doesn't recognise it as individual flags and gives an error.

**Reply by fdncred** [reply]
I think you can put the parameters into a list like this:
```
let fls = [($fl1) ($fl2)]
^$"ls" $fls
```

**Reply by yerlaser** [reply]
OK. I will try that.

**Reply by yerlaser** [reply]
Unfortunately, doesn't work:
```
╭───┬──────────────╮
│ 0 │ -p 8080:8080 │
│ 1 │ -p 8081:8081 │
│ 2 │ -p 8082:8082 │
╰───┴──────────────╯

Error: parsing host port: invalid port number: strconv.Atoi: parsing " 8080": invalid syntax
```

**Reply by fdncred** [reply]
This example with `ls` works, when I try it. I wonder what the difference is? 
```
let fl1 = '-F'
let fl2 = '-A'
let fls = [($fl1) ($fl2)]
^$"ls" $fls
```

The only thing I can see is that you're not just passing a flag. You're passing a flag and a value. You may have to put each of those items in the list separately. 

There are also some docker scripts in our nu_scripts repo. Not sure if they're helpful to this or not.

**Reply by yerlaser** [reply]
Yes, that's what I do. So, I need to create a two dimensional list where first item is always `-p` and second the ports, e.g. `8080:8080`? I'll try that later.

**Reply by fdncred** [reply]
I was thinking of something like this in nushell
```
["-p" "8080:8080" "-p" "8081:8081" "-p" "8082:8082"]
╭───┬───────────╮
│ 0 │ -p        │
│ 1 │ 8080:8080 │
│ 2 │ -p        │
│ 3 │ 8081:8081 │
│ 4 │ -p        │
│ 5 │ 8082:8082 │
╰───┴───────────╯
```
instead of this
```
╭───┬──────────────╮
│ 0 │ -p 8080:8080 │
│ 1 │ -p 8081:8081 │
│ 2 │ -p 8082:8082 │
╰───┴──────────────╯
```

**Reply by yerlaser** [reply]
Yes, that works, thanks!
But, I think it should be reflected somewhere in the docs because it's not very obvious.

**Reply by fdncred** [reply]
I wouldn't be surprised if our documentation was out of date. Feel free to write something up for it in our nushell.github.io repo.

**Comment by steelywing** [comment]
I use this

```nu
def "split command" [
    --quote (-q): string = "'" # the character of quote
] {
    return (
        ($in | from csv --no-infer --noheaders --separator ' ' --quote $quote).0 
        | values 
        | where { not ($in | is-empty) }
    )
}
```

then I can run command like

```nu
let $cmd = ("ffmpeg -i 'in video.mkv' -c copy 'out video.mkv'" | split command)
^$cmd.0 ...($cmd | skip 1)
```

Note that it only can use single type of quote, it can be specify with `--quote` parameter

**Comment by idoric** [comment]
At the moment when I am writing my answer (i.e. three years later) there’s now a built-in command designed for this use:

```
let compiler = "clang++"
let help_param = "--help"
run-external $compiler $help_param
```

This solution is particularly convenient for more complex situations :

```
let app = [flatpak run org.soundconverter.SoundConverter]
let param = [-b -m cbr -f mp3 ~/cd-wav/*.wav -o ~/cd-mp3]
run-external ...$app ...$param
```

**Reply by Mrfiregem** [reply]
FYI, `^` is shorthand for `run-external`, and the command it calls can be given as a list, so you can shorten it to `^$app ...$params` in your example (or technically if the command and it's params are all in one list, just `^$app`, but that's at the cost of readability imo)

**Reply by idoric** [reply]
I never imagined that `^$app ...$param` may work (and I don't really understand why), I had tried `^...$app ...$param` (and `^ ...$app ...$param`). That was the missing piece of my puzzle, thank you.
